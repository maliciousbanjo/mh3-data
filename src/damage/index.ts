export * as DamageTypes from './types';
export * as Util from './damage';
