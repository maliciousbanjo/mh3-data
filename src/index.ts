export * from './model';
export * as Damage from './damage';
