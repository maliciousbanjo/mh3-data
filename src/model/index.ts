export * as Common from './common';
export * as Items from './items';
export * as MonsterLevels from './monster-levels';
export * as Monsters from './monsters';
export * as Quests from './quests';
export * as Weapons from './weapons';
export * as ArmorSkills from './armor-skills';
