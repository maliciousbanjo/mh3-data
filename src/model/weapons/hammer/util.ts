import { MonsterLevelTypes } from '../../monster-levels';
import { MonsterTypes } from '../../monsters';
import { Attack, Damage, Sharpness, Weapon, WeaponType } from '../types';
import {
  calculateElementalDamage,
  getSharpnessRawMultiplier,
  isCutHit
} from '../util';
import { HammerDamageProperties } from './hammer-data';
import { HammerAttack } from './types';

function getHammerAttack(attackName: HammerAttack): Attack<HammerAttack> {
  const hammerAttacks = HammerDamageProperties.attackGroups[0];

  const result = hammerAttacks.attacks.find(atk => atk.name === attackName);
  if (!result) {
    throw new Error(`${attackName} is not a valid Hammer attack`);
  }
  return result;
}

/**
 * Calculates damage for a {@link HammerAttack}
 */
export function calculateHammerDamage(
  hammer: Weapon<WeaponType>,
  /** Attack being performed by the weapon */
  attackName: HammerAttack,
  /** Current sharpness of weapon */
  sharpness: Sharpness,
  /** Derived from Monster hitzone */
  hitzoneValues: MonsterTypes.HitzoneValues,
  /** Derived from Monster level set by Quest */
  multipliers: MonsterLevelTypes.MonsterLevelMultipliers,
  /** Will include awakened element if applicable */
  awaken = false
) {
  if (hammer.type !== WeaponType.HAMMER) {
    throw new Error(`${hammer.name} is not a ${WeaponType.HAMMER}`);
  }

  const { classModifier } = HammerDamageProperties;
  const sharpnessMultiplier = getSharpnessRawMultiplier(sharpness);

  const attack = getHammerAttack(attackName);

  // TODO: This should probably get lifted into a shared function that all weapons can use
  return attack.hits.map<Damage>(hit => {
    const isCut = isCutHit(hit);
    const hitzoneMultiplier = isCut ? hitzoneValues.cut : hitzoneValues.impact;

    const rawDamage =
      (hammer.attack *
        hit.power *
        sharpnessMultiplier *
        hitzoneMultiplier *
        1 * // Hammer does not have a [SpecialVar], but this is here for consistency
        multipliers.defense) /
      classModifier;

    const elementalDamage = calculateElementalDamage(
      hammer,
      sharpness,
      hitzoneValues,
      multipliers,
      awaken
    );

    const koDamage = !isCut ? hit.ko * sharpnessMultiplier : undefined;

    return {
      rawDamage,
      elementalDamage,
      totalDamage: Math.floor(rawDamage + elementalDamage),
      koDamage
    };
  });
}
