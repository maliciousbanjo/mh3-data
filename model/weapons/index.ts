// Data
import { GreatSwords, GreatSwordDamageProperties } from "./great-sword";
import {
  SwordAndShields,
  SwordAndShieldDamageProperties,
} from "./sword-and-shield";
import { Hammers, HammerDamageProperties } from "./hammer";
import { Lances, LanceDamageProperties} from './lance';

// Types
export type { GreatSwordTypes } from "./great-sword";
export type { SwordAndShieldTypes } from "./sword-and-shield";
export type { HammerTypes } from "./hammer";
export type { LanceTypes } from './lance';
export * from "./types";

export { GreatSwords, SwordAndShields, Hammers, Lances };

export const DamageProperties = {
  GreatSwordDamageProperties,
  SwordAndShieldDamageProperties,
  HammerDamageProperties,
  LanceDamageProperties
};
