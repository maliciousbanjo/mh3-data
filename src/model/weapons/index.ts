import type * as WeaponTypes from './types';
import type { GreatSwordTypes } from './great-sword';
import type { HammerTypes } from './hammer';
import type { LanceTypes } from './lance';
import type { LongswordTypes } from './longsword';
import type { SwitchAxeTypes } from './switch-axe';
import type { SwordAndShieldTypes } from './sword-and-shield';

import * as GreatSword from './great-sword';
import * as Hammer from './hammer';
import * as Lance from './lance';
import * as Longsword from './longsword';
import * as SwitchAxe from './switch-axe';
import * as SwordAndShield from './sword-and-shield';

export type {
  WeaponTypes,
  GreatSwordTypes,
  HammerTypes,
  LanceTypes,
  LongswordTypes,
  SwitchAxeTypes,
  SwordAndShieldTypes
};

export * from './enum';
export * from './weapon-util';

export { GreatSword, Hammer, Lance, Longsword, SwitchAxe, SwordAndShield };
