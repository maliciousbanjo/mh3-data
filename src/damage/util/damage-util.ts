import { WeaponTypes } from '../../model/weapons';
import { Weapons } from '../../model';
import { MonsterTypes } from '../../model/monsters';
import { MonsterLevelTypes } from '../../model/monster-levels';

const ELEMENTAL_DAMAGE_DIVIDER = 10;

/**
 * @returns true if weapon has the provided sharpness level, otherwise throws error
 * @throws Error if weapon does not have provided sharpness level
 */
export function validateWeaponSharpness(
  weapon: WeaponTypes.Weapon<WeaponTypes.WeaponClass>,
  sharpness: WeaponTypes.Sharpness
): true {
  const validSharpness = sharpness <= weapon.sharpnessUp.length - 1;
  if (!validSharpness) {
    throw new Error(
      `${weapon.name} cannot have ${Weapons.Util.sharpnessAsString(sharpness)} sharpness`
    );
  }
  return validSharpness;
}

/**
 * @returns sharpness multiplier for RAW damage (as opposed to ELEMENTAL)
 */
export function getSharpnessRawMultiplier(
  sharpness: WeaponTypes.Sharpness
): number {
  switch (sharpness) {
    case WeaponTypes.Sharpness.RED:
      return 0.5;
    case WeaponTypes.Sharpness.ORANGE:
      return 0.75;
    case WeaponTypes.Sharpness.YELLOW:
      return 1;
    case WeaponTypes.Sharpness.GREEN:
      return 1.05;
    case WeaponTypes.Sharpness.BLUE:
      return 1.2;
    case WeaponTypes.Sharpness.WHITE:
      return 1.32;
    case WeaponTypes.Sharpness.PURPLE:
      return 1.5;
    default:
      throw new Error(`Invalid sharpness value ${sharpness}`);
  }
}

/**
 * @returns sharpness multiplier for ELEMENTAL damage (as opposed to RAW)
 */
export function getSharpnessElementalMultiplier(
  sharpness: WeaponTypes.Sharpness
): number {
  switch (sharpness) {
    case WeaponTypes.Sharpness.RED:
      return 0.25;
    case WeaponTypes.Sharpness.ORANGE:
      return 0.55;
    case WeaponTypes.Sharpness.YELLOW:
      return 0.75;
    case WeaponTypes.Sharpness.GREEN:
      return 1;
    case WeaponTypes.Sharpness.BLUE:
      return 1.0625;
    case WeaponTypes.Sharpness.WHITE:
      return 1.125;
    case WeaponTypes.Sharpness.PURPLE:
      return 1.2;
    default:
      throw new Error(`Invalid sharpness value ${sharpness}`);
  }
}

/**
 * Used by {@link calculateElementalDamage}
 * @returns Elemental damage or 0 if weapon has status rather than element
 */
function getHitzoneForWeaponElement(
  hitzoneValues: MonsterTypes.HitzoneValues,
  weaponElement: WeaponTypes.Weapon<WeaponTypes.WeaponClass>['secondaryDamageType']
) {
  if (
    weaponElement === 'sleep' ||
    weaponElement === 'poison' ||
    weaponElement === 'paralysis'
  )
    return 0;

  switch (weaponElement) {
    case 'fire':
      return hitzoneValues.fire;
    case 'water':
      return hitzoneValues.water;
    case 'ice':
      return hitzoneValues.ice;
    case 'thunder':
      return hitzoneValues.thunder;
    case 'dragon':
      return hitzoneValues.dragon;
    default:
      throw new Error(`Invalid hitzone type ${weaponElement}`);
  }
}

/**
 * ELEMENTAL DAMAGE FORMULA
 *
 * Taken from Lord Grahf's [Monster Hunter Tri Damage Formula FAQ](https://gamefaqs.gamespot.com/wii/943655-monster-hunter-tri/faqs/59207)
 *
 * [ELEMENT x ESHARP x ELMZONE] / [DIVIDER] = Elemental Damage
 *
 * @example
 * [ELEMENT]:  250    // (250 thunder element)
 * [ESHARP]:   1.0    // (Green sharpness for elements, x 1.0)
 * [ELMZONE]:  .20    // (Rathian's weakness to thunder at head is 20)
 * [DIVIDER]:   10    // (Elemental Divider is always 10)
 * [DEFENSE]:  .75    // (.75 online high rank defense from earlier)
 *
 * [ELEMENT x ESHARP x ELMZONE] / [DIVIDER] = Elemental Damage [X DEFENSE]
 *    250   x  1.0   x   .20    /     10    =      5 (Added Thunder Damage)
 */
export function calculateElementalDamage(
  weapon: WeaponTypes.Weapon<WeaponTypes.ValidWeaponClasses>,
  /** Current sharpness of weapon */
  sharpness: WeaponTypes.Sharpness,
  /** Derived from Monster hitzone */
  hitzoneValues: MonsterTypes.HitzoneValues,
  multipliers: MonsterLevelTypes.MonsterLevelMultipliers,
  awaken = false
): number {
  const sharpnessMultiplier = getSharpnessElementalMultiplier(sharpness);
  const elementalHitzoneMultiplier = getHitzoneForWeaponElement(
    hitzoneValues,
    weapon.secondaryDamageType
  );

  // Element is not calculated for non-awakened weapon or if hitzone multiplier is 0
  if ((weapon.awaken && !awaken) || elementalHitzoneMultiplier === 0) return 0;

  return (
    (weapon.secondaryAttack *
      sharpnessMultiplier *
      elementalHitzoneMultiplier *
      multipliers.defense) /
    ELEMENTAL_DAMAGE_DIVIDER
  );
}
