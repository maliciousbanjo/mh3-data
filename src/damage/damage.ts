import { DamageBuffArgs, MonsterArgs, WeaponArgs } from './types';
import { Weapons } from '../model';
import { calculateGreatSwordDamage } from './util/great-sword-damage';
import { calculateHammerDamage } from './util/hammer-damage';
import { calculateLanceDamage } from './util/lance-damage';
import { calculateLongswordDamage } from './util/longsword-damage';
import { calculateSwitchAxeDamage } from './util/switch-axe-damage';
import { calculateSwordAndShieldDamage } from './util/sword-and-shield-damage';

/**
 * TODO: Affinity (positive or negative critical hit)
 * RAW DAMAGE FORMULA
 *
 * Taken from Lord Grahf's [Monster Hunter Tri Damage Formula FAQ](https://gamefaqs.gamespot.com/wii/943655-monster-hunter-tri/faqs/59207)
 *
 * [ATP x TYPE x SHARP x HITZONE x VAR] / [CLASS] = Raw Damage
 *
 * @example
 * [ATP]:     494     // (Lance 483 + attack boost 11)
 * [TYPE]:    .23     // (Normal lance stab, 23%)
 * [SHARP]:   1.05    // (Green sharpness 105%, or 1.05)
 * [HITZONE]: .90     // (Rathian head multiplier is 90)
 * [DEFENSE]: .75     // (In this high rank online quest, defense is .75)
 * [VAR]:     1.0     // (No special variable for lance)
 * [CLASS]:   2.3     // (All lances, class multiplier 2.3)
 *
 * [ATP x TYPE x SHARP x HITZONE] / [CLASS] = Raw Damage [X DEFENSE]
 *  494 x  .23 x 1.05  x   .90    /   2.3   =    46.683   X   .75
 */
export function calculateDamage(
  weaponArgs: WeaponArgs,
  monsterArgs: MonsterArgs,
  damageBuffArgs: DamageBuffArgs
) {
  const { weaponClass } = weaponArgs;
  switch (weaponClass) {
    case Weapons.WeaponTypes.WeaponClass.GREAT_SWORD: {
      return calculateGreatSwordDamage(weaponArgs, monsterArgs, damageBuffArgs);
    }
    case Weapons.WeaponTypes.WeaponClass.HAMMER: {
      return calculateHammerDamage(weaponArgs, monsterArgs, damageBuffArgs);
    }
    case Weapons.WeaponTypes.WeaponClass.LANCE: {
      return calculateLanceDamage(weaponArgs, monsterArgs, damageBuffArgs);
    }
    case Weapons.WeaponTypes.WeaponClass.LONGSWORD: {
      return calculateLongswordDamage(weaponArgs, monsterArgs, damageBuffArgs);
    }
    case Weapons.WeaponTypes.WeaponClass.SWITCH_AXE: {
      return calculateSwitchAxeDamage(weaponArgs, monsterArgs, damageBuffArgs);
    }
    case Weapons.WeaponTypes.WeaponClass.SWORD_AND_SHIELD: {
      return calculateSwordAndShieldDamage(
        weaponArgs,
        monsterArgs,
        damageBuffArgs
      );
    }
    default: {
      throw new Error(`${weaponClass} is not a valid weapon type`);
    }
  }
}
