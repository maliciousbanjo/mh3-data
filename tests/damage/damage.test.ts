import { calculateDamage } from '../../src/damage/damage';
import { validateWeaponSharpness } from '../../src/damage/util/damage-util';
import { MonsterLevels } from '../../src/model';
import {
  Alatreon,
  Deviljho,
  RoyalLudroth
} from '../../src/model/monsters/large-monster-data';
import { GreatSwords, WeaponTypes } from '../../src/model/weapons';

describe('Damage', () => {
  describe('damage utils', () => {
    describe('validateWeaponSharpness', () => {
      it('weapon has valid sharpness', () => {
        const gs = GreatSwords[0];
        expect(
          validateWeaponSharpness(gs, WeaponTypes.Sharpness.ORANGE)
        ).toBeTruthy();
      });
      it('throws error because weapon does not have sharpness', () => {
        const gs = GreatSwords[0];
        expect(() =>
          validateWeaponSharpness(gs, WeaponTypes.Sharpness.PURPLE)
        ).toThrow();
      });
    });
  });

  describe('calculateDamage', () => {
    const alatreonHeadHitzoneValues =
      Alatreon.hitzoneGroups[0].hitzones['Head'];
    const levelMultipliers = MonsterLevels.Util.getMonsterLevelMultipliers(52); // The Brilliant Darkness

    it('throws error for invalid sharpness', () => {
      expect(() =>
        calculateDamage(
          {
            weaponClass: WeaponTypes.WeaponClass.GREAT_SWORD,
            weaponId: 1, // Iron Blade
            attackName: 'L3 Charge',
            sharpness: WeaponTypes.Sharpness.PURPLE,
            weaponMultipliers: {}
          },
          { hitzoneValues: alatreonHeadHitzoneValues, levelMultipliers },
          {
            criticalHit: 'none',
            elementArgs: {
              awaken: false
            }
          }
        )
      ).toThrow();
    });

    it('Great Sword', () => {
      const damage = calculateDamage(
        {
          weaponClass: WeaponTypes.WeaponClass.GREAT_SWORD,
          weaponId: 57, // Anguish (P)
          attackName: 'L3 Charge',
          sharpness: WeaponTypes.Sharpness.PURPLE,
          weaponMultipliers: {
            middleOfBlade: false
          }
        },
        { hitzoneValues: alatreonHeadHitzoneValues, levelMultipliers },
        {
          criticalHit: 'none',
          elementArgs: {
            awaken: false
          }
        }
      );

      expect(damage).toMatchSnapshot();
    });

    it('Hammer', () => {
      const damage = calculateDamage(
        {
          weaponClass: WeaponTypes.WeaponClass.HAMMER,
          weaponId: 49, // Devil's Crush (P)
          attackName: 'Superpound',
          sharpness: WeaponTypes.Sharpness.PURPLE,
          weaponMultipliers: {}
        },
        { hitzoneValues: alatreonHeadHitzoneValues, levelMultipliers },
        {
          criticalHit: 'none',
          elementArgs: {
            awaken: false
          }
        }
      );
      expect(damage).toMatchSnapshot();
    });

    describe('Lance', () => {
      it('Cut hitzone', () => {
        const damage = calculateDamage(
          {
            weaponClass: WeaponTypes.WeaponClass.LANCE,
            weaponId: 52, // Alatreon Gleam,
            attackName: 'High Stab Combo',
            sharpness: WeaponTypes.Sharpness.PURPLE,
            weaponMultipliers: {}
          },
          { hitzoneValues: alatreonHeadHitzoneValues, levelMultipliers },
          {
            criticalHit: 'none',
            elementArgs: {
              awaken: false
            }
          }
        );
        expect(damage).toMatchSnapshot();
      });

      it('Modified impact hitzone', () => {
        // Royal Ludroth head (impact: 0.8, cut: 0.5), so the impact hitzone should be used
        const ludrothLevelMultipliers =
          MonsterLevels.Util.getMonsterLevelMultipliers(5); // Save Our Boat

        const royalLudrothHeadHitzoneValues =
          RoyalLudroth.hitzoneGroups[0].hitzones['Head'];
        const damage = calculateDamage(
          {
            weaponClass: WeaponTypes.WeaponClass.LANCE,
            weaponId: 52, // Alatreon Gleam,
            attackName: 'High Stab Combo',
            sharpness: WeaponTypes.Sharpness.PURPLE,
            weaponMultipliers: {}
          },
          {
            hitzoneValues: royalLudrothHeadHitzoneValues,
            levelMultipliers: ludrothLevelMultipliers
          },
          {
            criticalHit: 'none',
            elementArgs: {
              awaken: false
            }
          }
        );

        expect(damage).toMatchSnapshot();
      });
    });

    // TODO: These numbers are way off from StillTruth
    describe('Longsword', () => {
      it('throws error for missing special args', () => {
        expect(() =>
          calculateDamage(
            {
              weaponClass: WeaponTypes.WeaponClass.LONGSWORD,
              weaponId: 19, // Dark Claw Demise
              attackName: 'Spirit Finisher',
              sharpness: WeaponTypes.Sharpness.PURPLE,
              weaponMultipliers: {}
            },
            { hitzoneValues: alatreonHeadHitzoneValues, levelMultipliers },
            {
              criticalHit: 'none',
              elementArgs: {
                awaken: false
              }
            }
          )
        ).toThrow();
      });
      it('Spirit Gauge NONE', () => {
        const damage = calculateDamage(
          {
            weaponClass: WeaponTypes.WeaponClass.LONGSWORD,
            weaponId: 19, // Dark Claw Demise
            attackName: 'Spirit Finisher',
            sharpness: WeaponTypes.Sharpness.PURPLE,
            weaponMultipliers: {
              longsword: {
                fullSpiritGauge: true,
                spiritGaugeColor: 'NONE'
              }
            }
          },
          { hitzoneValues: alatreonHeadHitzoneValues, levelMultipliers },
          {
            criticalHit: 'none',
            elementArgs: {
              awaken: false
            }
          }
        );
        expect(damage).toMatchSnapshot();
      });
      it('Spirit Gauge WHITE', () => {
        const damage = calculateDamage(
          {
            weaponClass: WeaponTypes.WeaponClass.LONGSWORD,
            weaponId: 19, // Dark Claw Demise
            attackName: 'Spirit Finisher',
            sharpness: WeaponTypes.Sharpness.PURPLE,
            weaponMultipliers: {
              longsword: {
                fullSpiritGauge: true,
                spiritGaugeColor: 'WHITE'
              }
            }
          },
          { hitzoneValues: alatreonHeadHitzoneValues, levelMultipliers },
          {
            criticalHit: 'none',
            elementArgs: {
              awaken: false
            }
          }
        );
        expect(damage).toMatchSnapshot();
      });
      it('Spirit Gauge YELLOW', () => {
        const damage = calculateDamage(
          {
            weaponClass: WeaponTypes.WeaponClass.LONGSWORD,
            weaponId: 19, // Dark Claw Demise
            attackName: 'Spirit Finisher',
            sharpness: WeaponTypes.Sharpness.PURPLE,
            weaponMultipliers: {
              longsword: {
                fullSpiritGauge: true,
                spiritGaugeColor: 'YELLOW'
              }
            }
          },
          { hitzoneValues: alatreonHeadHitzoneValues, levelMultipliers },
          {
            criticalHit: 'none',
            elementArgs: {
              awaken: false
            }
          }
        );
        expect(damage).toMatchSnapshot();
      });
      it('Spirit Gauge RED', () => {
        const damage = calculateDamage(
          {
            weaponClass: WeaponTypes.WeaponClass.LONGSWORD,
            weaponId: 19, // Dark Claw Demise
            attackName: 'Spirit Finisher',
            sharpness: WeaponTypes.Sharpness.PURPLE,
            weaponMultipliers: {
              longsword: {
                fullSpiritGauge: true,
                spiritGaugeColor: 'RED'
              }
            }
          },
          { hitzoneValues: alatreonHeadHitzoneValues, levelMultipliers },
          {
            criticalHit: 'none',
            elementArgs: {
              awaken: false
            }
          }
        );
        expect(damage).toMatchSnapshot();
      });

      it('Deviljho Spirit Gauge RED', () => {
        // rage
        const deviljhoStomachHitzoneValues =
          Deviljho.hitzoneGroups[1].hitzones['Stomach'];
        const damage = calculateDamage(
          {
            weaponClass: WeaponTypes.WeaponClass.LONGSWORD,
            weaponId: 19, // Dark Claw Demise
            attackName: 'Step Slash',
            sharpness: WeaponTypes.Sharpness.PURPLE,
            weaponMultipliers: {
              longsword: {
                fullSpiritGauge: true,
                spiritGaugeColor: 'RED'
              }
            }
          },
          { hitzoneValues: deviljhoStomachHitzoneValues, levelMultipliers },
          {
            criticalHit: 'none',
            elementArgs: {
              awaken: false
            }
          }
        );
        expect(damage).toMatchSnapshot();
      });
    });

    describe('Switch Axe', () => {
      it('Throws error for missing special args', () => {
        expect(() =>
          calculateDamage(
            {
              weaponClass: WeaponTypes.WeaponClass.SWITCH_AXE,
              weaponId: 22, // Great Inceadeus
              attackName: 'Overhead Slash',
              sharpness: WeaponTypes.Sharpness.WHITE,
              weaponMultipliers: {}
            },
            { hitzoneValues: alatreonHeadHitzoneValues, levelMultipliers },
            {
              criticalHit: 'none',
              elementArgs: {
                awaken: false
              }
            }
          )
        ).toThrow();
      });
      it('Power phial', () => {
        const damage = calculateDamage(
          {
            weaponClass: WeaponTypes.WeaponClass.SWITCH_AXE,
            weaponId: 22, // Great Inceadeus
            attackName: 'Overhead Slash',
            sharpness: WeaponTypes.Sharpness.WHITE,
            weaponMultipliers: {
              switchAxeMode: 'sword'
            }
          },
          { hitzoneValues: alatreonHeadHitzoneValues, levelMultipliers },
          {
            criticalHit: 'none',
            elementArgs: {
              awaken: false
            }
          }
        );
        expect(damage).toMatchSnapshot();
      });
      it('Element phial', () => {
        const damage = calculateDamage(
          {
            weaponClass: WeaponTypes.WeaponClass.SWITCH_AXE,
            weaponId: 33, // Black Harvest
            attackName: 'Overhead Slash',
            sharpness: WeaponTypes.Sharpness.PURPLE,
            weaponMultipliers: {
              switchAxeMode: 'sword'
            }
          },
          { hitzoneValues: alatreonHeadHitzoneValues, levelMultipliers },
          {
            criticalHit: 'none',
            elementArgs: {
              awaken: false
            }
          }
        );
        expect(damage).toMatchSnapshot();
      });
    });

    describe('Sword and Shield', () => {
      it('Throws error for missing special args', () => {
        expect(() =>
          calculateDamage(
            {
              weaponClass: WeaponTypes.WeaponClass.SWORD_AND_SHIELD,
              weaponId: 55, // Alatreon Star
              attackName: 'Jumping Slash',
              sharpness: WeaponTypes.Sharpness.PURPLE,
              weaponMultipliers: {}
            },
            { hitzoneValues: alatreonHeadHitzoneValues, levelMultipliers },
            {
              criticalHit: 'none',
              elementArgs: {
                awaken: false
              }
            }
          )
        ).toThrow();
      });

      it('Cut attack', () => {
        const damage = calculateDamage(
          {
            weaponClass: WeaponTypes.WeaponClass.SWORD_AND_SHIELD,
            weaponId: 55, // Alatreon Star
            attackName: 'Jumping Slash',
            sharpness: WeaponTypes.Sharpness.PURPLE,
            weaponMultipliers: {
              swordAndShieldMode: 'land'
            }
          },
          { hitzoneValues: alatreonHeadHitzoneValues, levelMultipliers },
          {
            criticalHit: 'none',
            elementArgs: {
              awaken: false
            }
          }
        );

        expect(damage).toMatchSnapshot();
      });

      it('Impact attack', () => {
        const damage = calculateDamage(
          {
            weaponClass: WeaponTypes.WeaponClass.SWORD_AND_SHIELD,
            weaponId: 55, // Alatreon Star
            attackName: 'Shield Combo',
            sharpness: WeaponTypes.Sharpness.PURPLE,
            weaponMultipliers: {
              swordAndShieldMode: 'land'
            }
          },
          { hitzoneValues: alatreonHeadHitzoneValues, levelMultipliers },
          {
            criticalHit: 'none',
            elementArgs: {
              awaken: false
            }
          }
        );

        expect(damage).toMatchSnapshot();
      });
    });
  });
});
