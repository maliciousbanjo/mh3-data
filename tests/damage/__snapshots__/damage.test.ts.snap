// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Damage calculateDamage Great Sword 1`] = `
[
  {
    "elementalDamage": 0.675,
    "koDamage": undefined,
    "rawDamage": 144.78750000000002,
    "totalDamage": 145,
  },
]
`;

exports[`Damage calculateDamage Hammer 1`] = `
[
  {
    "elementalDamage": 0.675,
    "koDamage": 0,
    "rawDamage": 22.5,
    "totalDamage": 23,
  },
  {
    "elementalDamage": 0.675,
    "koDamage": 72,
    "rawDamage": 85.49999999999999,
    "totalDamage": 86,
  },
]
`;

exports[`Damage calculateDamage Lance Cut hitzone 1`] = `
[
  {
    "elementalDamage": 2.025,
    "koDamage": undefined,
    "rawDamage": 25.515000000000004,
    "totalDamage": 27,
  },
  {
    "elementalDamage": 2.025,
    "koDamage": undefined,
    "rawDamage": 25.515000000000004,
    "totalDamage": 27,
  },
  {
    "elementalDamage": 2.025,
    "koDamage": undefined,
    "rawDamage": 27.337500000000002,
    "totalDamage": 29,
  },
]
`;

exports[`Damage calculateDamage Lance Modified impact hitzone 1`] = `
[
  {
    "elementalDamage": 2.7,
    "koDamage": undefined,
    "rawDamage": 47.17440000000001,
    "totalDamage": 49,
  },
  {
    "elementalDamage": 2.7,
    "koDamage": undefined,
    "rawDamage": 47.17440000000001,
    "totalDamage": 49,
  },
  {
    "elementalDamage": 2.7,
    "koDamage": undefined,
    "rawDamage": 50.54400000000001,
    "totalDamage": 53,
  },
]
`;

exports[`Damage calculateDamage Longsword Deviljho Spirit Gauge RED 1`] = `
[
  {
    "elementalDamage": 14.804999999999998,
    "koDamage": undefined,
    "rawDamage": 66.63384,
    "totalDamage": 81,
  },
]
`;

exports[`Damage calculateDamage Longsword Spirit Gauge NONE 1`] = `
[
  {
    "elementalDamage": 2.115,
    "koDamage": undefined,
    "rawDamage": 37.06965,
    "totalDamage": 39,
  },
]
`;

exports[`Damage calculateDamage Longsword Spirit Gauge RED 1`] = `
[
  {
    "elementalDamage": 2.115,
    "koDamage": undefined,
    "rawDamage": 48.190545,
    "totalDamage": 50,
  },
]
`;

exports[`Damage calculateDamage Longsword Spirit Gauge WHITE 1`] = `
[
  {
    "elementalDamage": 2.115,
    "koDamage": undefined,
    "rawDamage": 38.9231325,
    "totalDamage": 41,
  },
]
`;

exports[`Damage calculateDamage Longsword Spirit Gauge YELLOW 1`] = `
[
  {
    "elementalDamage": 2.115,
    "koDamage": undefined,
    "rawDamage": 40.776615,
    "totalDamage": 42,
  },
]
`;

exports[`Damage calculateDamage Switch Axe Element phial 1`] = `
[
  {
    "elementalDamage": 2.0250000000000004,
    "koDamage": undefined,
    "rawDamage": 30.982499999999998,
    "totalDamage": 33,
  },
]
`;

exports[`Damage calculateDamage Switch Axe Power phial 1`] = `
[
  {
    "elementalDamage": 3.54375,
    "koDamage": undefined,
    "rawDamage": 35.974125,
    "totalDamage": 39,
  },
]
`;

exports[`Damage calculateDamage Sword and Shield Cut attack 1`] = `
[
  {
    "elementalDamage": 1.6650000000000003,
    "koDamage": undefined,
    "rawDamage": 15.4548,
    "totalDamage": 17,
  },
]
`;

exports[`Damage calculateDamage Sword and Shield Impact attack 1`] = `
[
  {
    "elementalDamage": 1.6650000000000003,
    "koDamage": 22.5,
    "rawDamage": 8.1,
    "totalDamage": 9,
  },
  {
    "elementalDamage": 1.6650000000000003,
    "koDamage": 22.5,
    "rawDamage": 12.149999999999999,
    "totalDamage": 13,
  },
]
`;
